File: app/src/main/java/com/cookplan/recipe/steps_mode/RecipeStepsInrgedientsAdapter.java
package com.cookplan.recipe.steps_mode;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.cookplan.R;
import com.cookplan.models.Ingredient;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by DariaEfimova on 18.03.17.
 */

public class RecipeStepsInrgedientsAdapter extends RecyclerView.Adapter<RecipeStepsInrgedientsAdapter.MainViewHolder> {

    private List<Ingredient> ingredients;
    private List<Ingredient> selectedIngredients;

    public RecipeStepsInrgedientsAdapter(List<Ingredient> ingredients) {
        this.ingredients = ingredients;
        selectedIngredients = new ArrayList<>();
    }

    @Override
    public MainViewHolder onCreateViewHolder(ViewGroup parent,
                                             int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.recipe_step_ingredients_item_layout, parent, false);

        MainViewHolder vh = new MainViewHolder(v);
        return vh;
    }

    @Override
    public void onBindViewHolder(MainViewHolder holder, int position) {
        Ingredient ingredient = ingredients.get(position);
        holder.nameTextView.setText(ingredient.getName());

        if (ingredient.getMainAmount() != null && ingredient.getMainMeasureUnit() != null) {
            holder.amountTextView.setVisibility(View.VISIBLE);

            holder.amountTextView.setText(ingredient.getMainMeasureUnit().toValueString(ingredient.getMainAmount()));

        } else {
            holder.amountTextView.setVisibility(View.GONE);
        }

        if (selectedIngredients.contains(ingredient)) {
            holder.mainView.setBackgroundResource(R.color.accent_color_light);
        } else {
            holder.mainView.setBackgroundResource(R.color.white);
        }

        View.OnClickListener clickListener = view -> {
            Ingredient selectIngredient = (Ingredient) view.getTag();
            if (selectedIngredients.contains(selectIngredient)) {
                selectedIngredients.remove(selectIngredient);
            } else {
                selectedIngredients.add(selectIngredient);
            }
            notifyDataSetChanged();
        };
        holder.mainView.setTag(ingredient);
        holder.mainView.setOnClickListener(clickListener);
    }

    @Override
    public int getItemCount() {
        return ingredients.size();
    }

    public static class MainViewHolder extends RecyclerView.ViewHolder {
        public TextView nameTextView;
        public TextView amountTextView;
        public View mainView;

        public MainViewHolder(View v) {
            super(v);
            nameTextView = (TextView) v.findViewById(R.id.ingredient_item_name);
            amountTextView = (TextView) v.findViewById(R.id.ingredient_item_amount);
            mainView = v.findViewById(R.id.main_view);
        }
    }
}

File: app/src/main/java/com/cookplan/recipe/steps_mode/RecipeStepsInrgedientsAdapter.kt
package com.cookplan.recipe.steps_mode

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.cookplan.R
import com.cookplan.models.Ingredient
import com.cookplan.utils.MeasureUnitUtils

/**
 * Created by DariaEfimova on 18.03.17.
 */

class RecipeStepsInrgedientsAdapter(private val ingredients: List<Ingredient>) : RecyclerView.Adapter<RecipeStepsInrgedientsAdapter.MainViewHolder>() {
    private val selectedIngred: MutableList<Ingredient> = mutableListOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MainViewHolder =
            MainViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.recipe_step_ingredients_item_layout, parent, false))

    override fun onBindViewHolder(holder: MainViewHolder, position: Int) {
        val item = ingredients[position]
        holder.nameTextView.text = item.name

        holder.amountTextView.text = MeasureUnitUtils.valueToString(
                item.mainMeasureUnit,
                item.mainAmount) { holder.amountTextView.context }
        holder.amountTextView.visibility = if (item.mainAmount > 0.0) View.VISIBLE else View.GONE

        holder.mainView.setBackgroundResource(
                if (selectedIngred.contains(item)) R.color.accent_color_light else R.color.white)

        with(holder.mainView) {
            tag = item
            setOnClickListener {
                if (selectedIngred.contains(item)) {
                    selectedIngred.remove(item)
                } else {
                    selectedIngred.add(item)
                }
                notifyDataSetChanged()
            }
        }
    }

    override fun getItemCount(): Int {
        return ingredients.size
    }

    class MainViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        var nameTextView: TextView
        var amountTextView: TextView
        var mainView: View

        init {
            nameTextView = v.findViewById<TextView>(R.id.ingredient_item_name)
            amountTextView = v.findViewById<TextView>(R.id.ingredient_item_amount)
            mainView = v.findViewById(R.id.main_view)
        }
    }
}